python/mlx/nn/layers/linear.py

```diff
@@ -26,20 +41,92 @@ class Linear(Module):
 
     def __init__(self, input_dims: int, output_dims: int, bias: bool = True):
         super().__init__()
-        scale = math.sqrt(1 / input_dims)
+        self.input_dims = input_dims
+        self.output_dims = output_dims
+        self.weight = mx.zeros((output_dims, input_dims))
+        if bias:
+            self.bias = mx.zeros((output_dims,))
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        scale = math.sqrt(1.0 / self.input_dims)
         self.weight = mx.random.uniform(
             low=-scale,
             high=scale,
-            shape=(output_dims, input_dims),
+            shape=(self.output_dims, self.input_dims),
         )
+        if "bias" in self:
+            self.bias = mx.random.uniform(
+                low=-scale,
+                high=scale,
+                shape=(self.output_dims,),
+            )
+
+    def _extra_repr(self):
+        return f"input_dims={self.input_dims}, output_dims={self.output_dims}, bias={'bias' in self}"
+
+    def __call__(self, input):
+        output = input @ self.weight.T
+        if "bias" in self:
+            output = output + self.bias
+        return output
+
+
+class Bilinear(Module):
+    r"""Applies a bilinear transformation to the input.
+
+    Concretely:
+
+    .. math::
+
+        y = input1^\top W input2 + b
```

Comment:
Also you need some kind of indexing on the y and W for this to make sense. Like:

$y_i = x_1^\top W_i x_2 + b_i$

Reply:
I'm not sure if we really need indexing because the Linear layer doesn't use indexing either. Let me know what you think

Reply:
Ok fair enough!

Reply:
On second thought, I think this computation is really different than the linear and subscripts make sense here. It makes it more clear that it's not a single bilinear matrix but $d$ bilinear matrices. For example, in the equation $y = x W$ the output $y$ has dimension $d$. But here the output $y = x_1^\top W x_2$ has dimension $1$ which is not right.

Reply:
Yeah, each multiplication results in a scalar, but I think its similar to `Linear` because it corresponding to a `d` linear "vectors". In any case, I've changed the math formula to include 'i' in the latest commit  https://github.com/ml-explore/mlx/pull/315/commits/a9512835611dbaee74f567d9c6d2add34c0eb306.